var/spack/repos/builtin/packages/grpc/package.py

```diff
@@ -19,6 +19,8 @@ class Grpc(CMakePackage):
     version('1.24.3', sha256='c84b3fa140fcd6cce79b3f9de6357c5733a0071e04ca4e65ba5f8d306f10f033')
     version('1.23.1', sha256='dd7da002b15641e4841f20a1f3eb1e359edb69d5ccf8ac64c362823b05f523d9')
 
+    variant('shared', default=True,
```

Comment:
Changing the default could break lots of stuff downstream, would you mind switching the default to `False` to match the current behavior?

Reply:
First: I can change the default, if that's what's needed.

Points:
* Most other packages I am currently aware of have a default of shared library instead of static library. So I'd prefer that. Even though it might be a small breaking change.
* Could you elaborate on a use case that might break?
  * Most (all?) build systems should gracefully handle shared libraries instead of static ones.
  * If people want to build a static executable, they most likely have a global `~shared` setting, because of all the other libraries.
  * The possibly remaining use cases should be quite minor and special. Those users most likely "know, what they're doing".

Reply:
> Most other packages I am currently aware of have a default of shared library instead of static library.
Yes, that could be definitely the case for Spack packages but the default for this one has always been static. I would be totally cool with shared objs but the point is that switching defaults is a matter of Spack policies I guess, and I'm not a maintainer so I think we should ping @adamjstewart 

> Could you elaborate on a use case that might break?

Everyone that passes archives to the linker, for example. That's how CIs break and people blame package managers.

Reply:
* IMHO build systems that pass the archive directly to the linker instead of `-L{path} -lgrpc` are broken and should be fixed. (Fixed by either fixing the build system or explicitly doing `depends_on('grpc~shared')` in spack.)
* Yes, let's wait for adamjstewart's comment.

Reply:
Repetita iuvant:

> That's how CIs break and people end up blaming package managers.

Reply:
@ChristianTackeGSI Just my two cents, but I would be for keeping a non-breaking behavior for people that just:
```console
$ spack install grpc
```
unless there are clear reasons why a breaking behavior is preferable. Arguments like:

> If people want to build a static executable, they most likely have a global ~shared setting, because of all the other libraries.

seems more a logical fallacy to me, since the opposite:

> If people want to build a shared executable, they most likely have a global +shared setting, because of all the other libraries.

is equally true and just means that for those people the default doesn't matter.

Reply:
I personally like switching defaults because it breaks things. Most packages depend on other packages with no additional constraints. By switching defaults, you find out those additional constraints by breaking builds.

Reply:
@adamjstewart Either outcomes for this PR are fine with me, but I think that:

> I personally like switching defaults because it breaks things [...] you find out those additional constraints by breaking builds.

wouldn't apply much in this case since there's a single package that depends on `grpc`. I think the impact is more on people that for their use:
```console
$ spack install grpc
```
and would see a shared library in whatever environment they're building instead of a static one.

Reply:
That said my policy is that the maintainer of a package has the last word on semantic changes, so if it is fine with @nazavode it's green light also for me.

Reply:
> That said my policy is that the maintainer of a package has the last word on semantic changes, so if it is fine with @nazavode it's green light also for me.

Considering `grpc` just in the context of Spack (like for the point raised by @adamjstewart) this PR is fine as no packages depends on it and the change doesn't affect the functionalities provided by `grpc`, both the compiler plugin and the libraries. On the other hand, considering `grpc` used in the wild and deployed via Spack, this change would break everyone relying on the presence of archives, so I would prefer to avoid it if not necessary since I tend to see Spack as a tool suitable for production use. In summary: for the package maintainer is a no, for Spack as a whole is fine, looks like a deadlock :)

Reply:
Can we build both shared and static libraries? I think that's the default for most packages.

Reply:
> Can we build both shared and static libraries? I think that's the default for most packages.

Never tried myself but I think it's doable as long as the compiler plugin knows which one to pick. @ChristianTackeGSI would you mind giving it a try?

Reply:
cmake isn't great at building both at the same time, really. At least AFAIK. (Not a cmake expert).

If this makes things easier for everyone, I am happy to just continue building static libraries as before and thus only giving people who need it, the option to build shared libraries?